api/server.go

```diff
@@ -286,3 +295,85 @@ func loggingUnaryInterceptor(logger *zerolog.Logger) grpc.UnaryServerInterceptor
 		return resp, err
 	}
 }
+
+func (s *service) DetailedHealthCheck(ctx context.Context, req *task.DetailedHealthCheckRequest) (*task.DetailedHealthCheckResponse, error) {
+	var unhealthyReasons []string
+	resp := &task.DetailedHealthCheckResponse{}
+
+	criuVersion, err := s.CRIU.GetCriuVersion()
+	if err != nil {
+		resp.UnhealthyReasons = append(unhealthyReasons, fmt.Sprintf("CRIU: %v", err))
+	}
+
+	resp.HealthCheckStats = &task.HealthCheckStats{}
+	resp.HealthCheckStats.CriuVersion = strconv.Itoa(criuVersion)
+
+	// TODO NR - Add CRIU check to output
+	err = s.GPUHealthCheck(ctx, req, resp)
+	if err != nil {
+		resp.UnhealthyReasons = append(unhealthyReasons, fmt.Sprintf("Error checking gpu health: %v", err))
+	}
+
+	return resp, nil
+}
+
+func (s *service) GPUHealthCheck(
+	ctx context.Context,
+	req *task.DetailedHealthCheckRequest,
+	resp *task.DetailedHealthCheckResponse) error {
+	gpuControllerPath := viper.GetString("gpu_controller_path")
+	if gpuControllerPath == "" {
+		gpuControllerPath = utils.GpuControllerBinaryPath
+	}
+
+	gpuSharedLibPath := viper.GetString("gpu_shared_lib_path")
+	if gpuSharedLibPath == "" {
+		gpuSharedLibPath = utils.GpuSharedLibPath
+	}
+
+	if _, err := os.Stat(gpuControllerPath); os.IsNotExist(err) {
+		resp.UnhealthyReasons = append(resp.UnhealthyReasons, fmt.Sprintf("gpu controller binary not found at %s", gpuControllerPath))
+	}
+
+	if _, err := os.Stat(gpuSharedLibPath); os.IsNotExist(err) {
+		resp.UnhealthyReasons = append(resp.UnhealthyReasons, fmt.Sprintf("gpu shared lib not found at %s", gpuSharedLibPath))
+	}
+
+	if resp.UnhealthyReasons != nil {
+		return fmt.Errorf("GPU health check failed")
+	}
+
+	cmd, err := StartGPUController(ctx, req.UID, req.GID, req.Groups, s.logger)
+	if err != nil {
+		return err
+	}
+
+	defer cmd.Process.Kill()
+
+	var opts []grpc.DialOption
+	opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	gpuConn, err := grpc.Dial("127.0.0.1:50051", opts...)
+	if err != nil {
+		return err
```

Comment:
Same as above. Let's use error return only if the health check itself has a problem.